### PR TITLE
Feature/improve testing, resolves #5 and #1

### LINKED ISSUE #5: Improve testing
This is related to the [JOSS review](https://github.com/openjournals/joss-reviews/issues/8311).

I see that you are started creating a test-suite, which is a good start. However, this isn't using any industry standard tools, like `pytest` and is rather adhoc. I would recommend [reviewing the PyOpenSci testing guide](https://www.pyopensci.org/python-package-guide/tests/index.html). Right now it seems like you have around 1% test coverage? (see [attached coverage reports](https://github.com/user-attachments/files/20951891/htmlcov.zip)). 

It seems like your test is trying to do a large integration test and handle logging and reporting. A testing framework, like `pytest`, can handle the logging and reporting for you. In general [unit tests are much more effective than integration testing](https://dl.acm.org/doi/10.1145/2896941.2896944)

I would recommend:

* [x] implement `pytest` as a testing framework.
* [x] implement #1, so that you can ensure that you are testing the local code, and not the installed code
* [x] create unit tests to achieve 80+% coverage
* [x] Focus testing a lot on "physics related methods", i.e., those that are calculating new tallies, combining tallies, etc.; really the functions where the math has an impact on the end user.

### PR DESCRIPTION
**Pull request by Lindt8**
This PR implements testing, completely overhauling the previous approach.  Testing was developed with [`pytest`](https://pypi.org/project/pytest/) and [`pytest-cov`](https://pypi.org/project/pytest-cov/) (for coverage testing). Configuration details can be found in [`pyproject.toml`](https://github.com/Lindt8/PHITS-Tools/tree/feature/improve-testing/pyproject.toml) and common test execution commands in the [`Makefile`](https://github.com/Lindt8/PHITS-Tools/tree/feature/improve-testing/Makefile). 

The testing for PHITS Tools is divided into two directories:

- [`tests/`](https://github.com/Lindt8/PHITS-Tools/tree/feature/improve-testing/tests/) for **unit tests**: These are used for [CI testing](https://github.com/Lindt8/PHITS-Tools/actions/workflows/ci-tests.yml) and are automatically ran with pull requests to the `develop` and `main` branches.  This set of unit tests primarily is for the standalone functions in PHITS Tools that end-users are likely to use for additional results post-processing or independent use, rather than the functions which are complex internal cogs in output file processing (though, those which are suited for unit testing are included here too).  While only attaining ~19% total coverage, these tests achieve at least 90% coverage for all of the functions that are explicitly unit tested here.
- [`test/`](https://github.com/Lindt8/PHITS-Tools/tree/feature/improve-testing/test/) for **integration/functional tests**: These require externally distributed PHITS output files (see [`test/README.md`](https://github.com/Lindt8/PHITS-Tools/tree/feature/improve-testing/test/README.md) for full details) and test the start-to-finish function of PHITS Tools. These tests exist to verify if the code runs without error and, where reasonable, checks if the output has properties/formatting as expected.  These tests combined with the unit tests should achieve ~89% total coverage, with over 80% coverage in every function in the main PHITS Tools module (except the CLI and GUI launching functions, which are excluded).

Coverage reports are also uploaded to [Codecov](https://app.codecov.io/github/lindt8/phits-tools); automatically [via CI](https://github.com/Lindt8/PHITS-Tools/actions/workflows/ci-tests.yml) for the unit tests (see flag [`ci-unittests`](https://app.codecov.io/github/lindt8/phits-tools?flags%5B0%5D=ci-unittests)) and manually via the `make upload-full` command (which makes use of [`codecovcli`](https://pypi.org/project/codecov-cli/)) for the integration/functional tests (see flag [`full-suite`](https://app.codecov.io/github/lindt8/phits-tools?flags%5B0%5D=full-suite)), of course after running those tests with `make test-all`.

Resolves #5 and #1

### CODE REVIEW

File: PHITS_tools/PHITS_tools.py [hunk 1]
```diff
@@ -276,7 +276,7 @@
 #test_explicit_files_dirs = True
 
 
-if __name__ == "__main__":
+if __name__ == "__main__":   # pragma: no cover
     #in_debug_mode = True
```
[the remark this answers is not in the document]
Comment (MicahGale) [reply PHITS_tools/PHITS_tools.py:280]: Also that's one way to get your coverage up ;)

File: README.md [hunk 2]
```diff
@@ -203,10 +203,20 @@ these should work and be accessed as shown without any additional configuration
 
 ## Testing, reporting issues, and contributing
 
+[![CI Tests](https://github.com/Lindt8/PHITS-Tools/actions/workflows/ci-tests.yml/badge.svg)](https://github.com/Lindt8/PHITS-Tools/actions/workflows/ci-tests.yml)
+ ° [![](https://img.shields.io/badge/Unit%20tests%20only-gray) ![Unit Tests](https://codecov.io/gh/Lindt8/PHITS-Tools/branch/feature/improve-testing/graph/badge.svg?flag=ci-unittests&label=Unit%20Tests)](https://app.codecov.io/github/lindt8/phits-tools?flags%5B0%5D=ci-unittests)
```
Comment (MicahGale) [inline README.md:207]: This seems to be broken markdown, and the badges aren't all showing up.

Also don't be rude, and not include alt-text, one of the most fundamental parts of accessible web design.

File: codecov.yml
Comment (MicahGale) [inline codecov.yml:1]: Are you allowed to name this `.codecov.yml` and make it a hidden file?

File: pyproject.toml [hunk 3]
```diff
@@ -93,3 +95,80 @@ exclude = [
...
+minversion = "6.0"
+# Add useful command line options by default
+addopts = [
+    "--strict-markers",  # Treat unregistered markers as errors
+    "--strict-config",   # Treat config issues as errors
+    "--verbose",         # Verbose output
+    "-ra",              # Show extra test summary for all except passed
+#    "--cov=PHITS_tools", # Enable coverage by default
+#    "--cov-report=html", # Generate HTML report by default
+#    "--cov-report=term-missing", # Also show terminal report
```
Comment (MicahGale) [inline pyproject.toml:126]: Dead code.

File: pyproject.toml [hunk 4]
```diff
@@ -93,3 +95,80 @@ exclude = [
...
+]
+# Filter warnings 
+filterwarnings = [
+    "error",  # Treat warnings as errors (remove if too strict)
+    "ignore::UserWarning",
+    "ignore::DeprecationWarning",
+    "ignore::FutureWarning:PHITS_tools.*", # Only ignore from PHITS_tools module
```
Comment (MicahGale) [inline pyproject.toml:133]: Why are you doing this?

I think generally for warnings you should test that the warnings are raised:

``` python
with pytest.warns("...."):
    pass
```

File: pyproject.toml [hunk 5]
```diff
@@ -93,3 +95,80 @@ exclude = [
...
+source = ["PHITS_tools"]
+branch = true
+# Omit test files from coverage
+omit = [
+    "tests/*",
+    "test/*",
+    "*/tests/*",
+    "*/test/*",
+    "build_docs.py",
+]
```
Comment (MicahGale) [inline pyproject.toml:146]: I think this is redundant due to the use of `source`

File: pyproject.toml [hunk 6]
```diff
@@ -93,3 +95,80 @@ exclude = [
...
+# Filter warnings 
+filterwarnings = [
+    "error",  # Treat warnings as errors (remove if too strict)
+    "ignore::UserWarning",
+    "ignore::DeprecationWarning",
+    "ignore::FutureWarning:PHITS_tools.*", # Only ignore from PHITS_tools module
+]
+
+[tool.coverage.run]
+source = ["PHITS_tools"]
```
Comment (MicahGale) [inline pyproject.toml:137]: Shouldn't you also include `DCHAIN-tools` (or add its own test suite in repo), and `MC_materials`?

File: pyproject.toml [hunk 7]
```diff
@@ -93,3 +95,80 @@ exclude = [
...
+show_missing = true
+skip_covered = false
+skip_empty = true
+precision = 2
+exclude_lines = [
+    "pragma: no cover",
+    "if __name__ == '__main__':",
+    "if False:",
+    "if in_debug_mode:",
+    "raise ValueError",
```
Comment (MicahGale) [inline pyproject.toml:159]: I disagree here. I have had many new errors raised while trying to raise a `ValueError`.

File: pyproject.toml [hunk 8]
```diff
@@ -93,3 +95,80 @@ exclude = [
...
+]
+# Set coverage thresholds
+# Note: This is overridden in CI with --cov-fail-under=0
+# Local development can use full integration tests (89% coverage)
+fail_under = 80
+
+[tool.coverage.html]
+directory = "htmlcov"
```
Comment (MicahGale) [inline pyproject.toml:167]: `default` I think this is redundant.

File: pyproject.toml [hunk 9]
```diff
@@ -93,3 +95,80 @@ exclude = [
...
+title = "PHITS-Tools Test Coverage"
+skip_covered = false
+skip_empty = false
+#show_contexts = true
```
Comment (MicahGale) [inline pyproject.toml:171]: The code is more readable IMO in these reports if you get the full context.

File: pyproject.toml [hunk 10]
```diff
@@ -93,3 +95,80 @@ exclude = [
...
+
+[tool.coverage.xml]
+output = "coverage.xml"
```
Comment (MicahGale) [inline pyproject.toml:174]: This is also the default.

File: test/test_input_mode_parsing.py [hunk 11]
```diff
@@ -0,0 +1,96 @@
+import pytest, PHITS_tools
+from pathlib import Path
+from traceback import format_exc
+import re
+
+path_to_phits_base_folder = Path(r'C:\phits')
```
Comment (MicahGale) [inline test/test_input_mode_parsing.py:6]: I would use a environment variable to set this (maybe default to this if not set) so that linux developers can rule the world.

File: test/test_input_mode_parsing.py [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+
+phits_sample_dir = Path(path_to_phits_base_folder,'sample')
+phits_recommendation_dir = Path(path_to_phits_base_folder,'recommendation')
+test_autoplotting = True  # Determines if autoplot_tally_results() is tested too for each tally output (notably slows testing)
+known_issue_files = [r'phits\sample\source\Cosmicray\GCR-ground\cross.out', r'phits\sample\source\Cosmicray\GCR-ground\GCR-ground.inp',
+                     r'phits\sample\source\Cosmicray\GCR-LEO\cross.out', r'phits\sample\source\Cosmicray\GCR-LEO\GCR-LEO.inp',
+                     r'phits\sample\source\Cosmicray\GCR-space\cross.out', r'phits\sample\source\Cosmicray\GCR-space\GCR-space.inp',
+                     r'phits\sample\source\Cosmicray\SEP-space\cross.out', r'phits\sample\source\Cosmicray\SEP-space\SEP-space.inp',
+                     r'phits\sample\source\Cosmicray\TP-LEO\cross.out', r'phits\sample\source\Cosmicray\TP-LEO\TP-LEO.inp',
+                     r'phits\sample\misc\snowman\rotate3dshow.inp']
```
Comment (MicahGale) [inline test/test_input_mode_parsing.py:16]: Is this an issue with your software or PHITS?

Comment (MicahGale) [reply test/test_input_mode_parsing.py:16]: If your software; I'm not sure I like this. Maybe instead you could use `@pytest.expectedfailure`?

File: test/test_input_mode_parsing.py [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+
+potential_inputs_to_parse = [Path(f) for f in phits_sample_dir.rglob('*.inp')] + [Path(f) for f in phits_recommendation_dir.rglob('*.inp')]
+inputs_to_parse = []
+for f in potential_inputs_to_parse:
+    if 'benchmark' in str(f): continue  # this directory has lots of problem folders with missing outputs
+    if len([i for i in f.parent.glob('*.out')]) == 0: continue  # skip inputs in directories with no output files
+    skip_this_input = False
+    inputs_to_parse.append(f)
+numi_tests = len(inputs_to_parse)
+#log_str = '\n\n--------------------------------------------------------------------\n'
```
Comment (MicahGale) [inline test/test_input_mode_parsing.py:35]: These are a lot of messy global variables. I think `pytest.fixture` would be more appropriate.

File: test/test_input_mode_parsing.py [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+#log_str += 'INPUT FILE PARSING TESTS\n\n'
+#log_file_str += log_str
+
+# save log file, deleting any old contents
+with open(log_file_path, "w") as f:
+    f.write(log_file_str)
+
```
Comment (MicahGale) [inline test/test_input_mode_parsing.py:42]: This could lead to poluting the devs compute with extra files, and cause weird test states. Could you: use `io.StringIO` instead? If not, do something like:

``` python
@pytest.fixture
def log_file():
    try:
         with open(log_file_path, "w") as f:
              f.write(log_file_str)
    finally:
         try:
            os.remove(log_file_path)
            yield
         except FileNotFoundError:
            pass
```
This would ensure the log file is always deleted.

Comment (MicahGale) [reply test/test_input_mode_parsing.py:42]: Probably want a global scope of the fixture so you don't write and delete the file 20 times.

File: test/test_input_mode_parsing.py [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+
+def process_and_log_output_from_input_file(f):
+    global log_str, log_file_str, i, numi_passed, numi_failed, numi_warn
```
Comment (MicahGale) [inline test/test_input_mode_parsing.py:45]: fixtures.

File: .github/workflows/ci-tests.yml [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+      uses: actions/setup-python@v4
+      with:
+        python-version: ${{ matrix.python-version }}
+
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install -e ".[develop]"
+
+    - name: Check for critical syntax errors
```
[the remark this answers is not in the document]
Comment (Lindt8) [reply .github/workflows/ci-tests.yml]: Yep, I've made it a separate job and only using one version of python for it now.

File: CONTRIBUTING.md [hunk 17]
```diff
@@ -17,20 +17,31 @@ If you would like to contribute a new function or changes to any existing functi
...
+
+Testing was developed with [`pytest`](https://pypi.org/project/pytest/) and [`pytest-cov`](https://pypi.org/project/pytest-cov/) (for coverage testing).  Configuration details can be found in [`pyproject.toml`](pyproject.toml) and common test execution commands in the [`Makefile`](Makefile).  Coverage reports are also uploaded to [Codecov](https://app.codecov.io/github/lindt8/phits-tools); automatically [via CI](https://github.com/Lindt8/PHITS-Tools/actions/workflows/ci-tests.yml) for the unit tests (see flag [`ci-unittests`](https://app.codecov.io/github/lindt8/phits-tools?flags%5B0%5D=ci-unittests)) and manually via the `make upload-full` command (which makes use of [`codecovcli`](https://pypi.org/project/codecov-cli/)) for the integration/functional tests (see flag [`full-suite`](https://app.codecov.io/github/lindt8/phits-tools?flags%5B0%5D=full-suite)), of course after running those tests with `make test-all`.
```
[the remark this answers is not in the document]
Comment (Lindt8) [reply CONTRIBUTING.md]: I've added explicit mention here of `make test` (run unit tests), `make test-cov` (coverage for unit tests), and `make test-all` (coverage for full test suite). I don't think it's necessary to maintain a separate full list of the make commands outside of the Makefile itself as long as those in the Makefile are adequately obvious and/or described in comments.

File: PHITS_tools/PHITS_tools.py [hunk 18]
```diff
@@ -984,17 +984,17 @@ def handle_dtrk_dyld_file(dtrk_dyld_filepath, tdchain_metadata={}):
...
         if err_data_found:
             if in_debug_mode: print("\nCalculating absolute errors...   ({:0.2f} seconds elapsed)".format(time.time() - start))
             tally_data = calculate_tally_absolute_errors(tally_data)
             if in_debug_mode: print("\tComplete!   ({:0.2f} seconds elapsed)".format(time.time() - start))
-        elif is_err_file:
+        elif is_err_file:  # pragma: no cover
             print('\tWARNING: Absolute errors not calculated since the main tally values file was not found.')
-        else:
+        else:  # pragma: no cover
             print('\tWARNING: Absolute errors not calculated since the _err file was not found.')
```
[the remark this answers is not in the document]
Comment (Lindt8) [reply PHITS_tools/PHITS_tools.py:998]: I have coded _a lot_ of contingencies into PHITS Tools, and it's really not feasible to test every single possible state.  In this specific example, these conditions are only ever met if the user has deleted or moved one of two linked output files but selected options requiring both to be found (in PHITS, if you specify a 2-dimensional output axis, values and relative errors are written to separate files; this is a quirk coming from the built-in EPS output visualization code in PHITS).  

Should this ever actually happen in practice?  Nope, not ever.  However, if a user did accidentally delete one of these files, would it be better to warn them of this and carry on (this is what I've opted for), just have the code crash without explanation, or have the code carry on without informing the user of this and that their output will differ from what they expected?

The code has a lot of these "What if ...?" contingencies I'd thought up while developing that I think are unlikely to really happen in practice, but on the off chance it did happen it'd be nice to issue a warning to the user.  So, similar to how [lines with `"raise ValueError"` are excluded from coverage testing](https://github.com/Lindt8/PHITS-Tools/blob/feature/improve-testing/pyproject.toml#L159), I've also excluded these edge-case warnings.

On top of this, there are other very unlikely conditions I wanted to pre-program some sort of behavior for even if they'd never occur with testing.  The plotting function has some of these.  Is it likely for you to ever have a scenario where you're trying to represent six different variables on the same plot?  Or is this even a good idea? Probably not to both, but can seaborn's relplot handle this? Yep.  So, why not include that functionality just on the off-chance some user ever triggers it?  But nothing in the PHITS-distributed outputs comes even close to triggering this, and it does seem a bit pointless to add extra complication to the testing procedure just to cover extreme outlier edge cases.

There are also other contingencies in place to handle some rare bugs/weird behaviors within the PHITS code itself that can be difficult to replicate (namely, in some cases where weird things happen in Fortran's write process), or rather extreme situations (like handling very large binary dump files tens of GB in size or larger, too big to have in memory all at once).

I have made sure any code I'd expect to somehow get triggered within 99% of various use-cases is included in the coverage testing.  However, I've chosen to exclude the contingencies I think are extremely rare from coverage testing.  This seems fair to me, as I think its preferable for PHITS Tools to have acknowledged and prepared for some weird edge cases but lack coverage testing that'd require a lot of extra complexity in the testing process with deleting/corrupting output files or having to generate some really wonky tally conditions in PHITS just to achieve a higher coverage metric.

File: README.md [hunk 19]
```diff
@@ -203,10 +203,20 @@ these should work and be accessed as shown without any additional configuration
...
+ ° [![](https://img.shields.io/badge/Unit%20tests%20only-gray) ![Unit Tests](https://codecov.io/gh/Lindt8/PHITS-Tools/branch/feature/improve-testing/graph/badge.svg?flag=ci-unittests&label=Unit%20Tests)](https://app.codecov.io/github/lindt8/phits-tools?flags%5B0%5D=ci-unittests) 
+ ° [![](https://img.shields.io/badge/Full%20test%20suite-gray) ![Full Suite](https://codecov.io/gh/Lindt8/PHITS-Tools/branch/feature/improve-testing/graph/badge.svg?flag=full-suite&label=Full%20Suite)](https://app.codecov.io/github/lindt8/phits-tools?flags%5B0%5D=full-suite)
+<!--
+ ° [![](https://img.shields.io/badge/Unit%20tests%20only-gray) ![Unit Tests](https://codecov.io/gh/Lindt8/PHITS-Tools/graph/badge.svg?flag=ci-unittests&label=Unit%20Tests)](https://app.codecov.io/github/lindt8/phits-tools?flags%5B0%5D=ci-unittests)
+ ° [![](https://img.shields.io/badge/Full%20test%20suite-gray) ![Full Suite](https://codecov.io/gh/Lindt8/PHITS-Tools/graph/badge.svg?flag=full-suite&label=Full%20Suite)](https://app.codecov.io/github/lindt8/phits-tools?flags%5B0%5D=full-suite)
+/-->
```
[the remark this answers is not in the document]
Comment (Lindt8) [reply README.md:212]: The active badges right now point specifically to this `feature/improve-testing` branch.  Once this code gets pushed to `main`, I should be able to replace the active badges with these commented-out badges.  (At present, they just show "unknown".)

File: codecov.yml
Comment (Lindt8) [reply codecov.yml:1]: I think so?  The documentation isn't super clear on it, but some googling makes me think it should work.  I actually don't think this is doing anything yet since it's not been pushed to the default branch, but I'll rename it and we'll see.

File: test/test_input_mode_parsing.py (hunk 12 above)
Comment (Lindt8) [reply test/test_input_mode_parsing.py:16]: It's an issue with the files in the PHITS distribution.  An [appropriate message is included in the log file](https://github.com/Lindt8/PHITS-Tools/blob/feature/improve-testing/test/test_overall_functionality.py#L91) if an issue with these files is encountered telling the user to just rerun the PHITS inputs in that folder to regenerate those outputs, after which the test should pass without issue.

File: test/test_input_mode_parsing.py (hunk 13 above)
Comment (Lindt8) [reply test/test_input_mode_parsing.py]: I've cleaned this up in 40c89d05e498900ef0bfc932c899072cd8b7e14b.

File: test/test_input_mode_parsing.py (hunk 14 above)
Comment (Lindt8) [reply test/test_input_mode_parsing.py]: I've cleaned this up in 40c89d05e498900ef0bfc932c899072cd8b7e14b, now using `io.StringIO`.  I am still saving the log file once at the very end though, but this is much better than repeatedly appending to it.

File: test/test_input_mode_parsing.py (hunk 15 above)
Comment (Lindt8) [reply test/test_input_mode_parsing.py]: Fixed in 40c89d05e498900ef0bfc932c899072cd8b7e14b.

**Comment by MicahGale** [comment]
I couldn't find any tests of `ICRP116_effective_dose_coeff`

File: test/test_options_in_processing.py [hunk 20]
```diff
@@ -0,0 +1,454 @@
...
+from pathlib import Path
+from conftest import path_to_base_phits_test_dir
+
+phits_sample_dir = Path(path_to_base_phits_test_dir,'sample')
+phits_recommendation_dir = Path(path_to_base_phits_test_dir,'recommendation')
+
+def test_extract_tally_outputs_from_phits_input_options():
+    phits_input = phits_recommendation_dir / 'NeutronSource' / 'NeutronSource.inp'
+    x = PHITS_tools.extract_tally_outputs_from_phits_input(phits_input, only_seek_phitsout=True)
+    assert x['phitsout'] != []
```
Comment (MicahGale) [inline test/test_options_in_processing.py:13]: You should also test that it gets the right answer, not just an answer.

**Review by MicahGale — CHANGES_REQUESTED** [review]
In general the tests need to be more concrete with what the expected output is not just that it gave an output.

**Comment by junghans** [comment]
Again this request seems to go a bit beyond the JOSS review, but as you are close we should wrap this up and leave the rest for after.
